File: ios-offline-maps/ios-offline-maps/OfflineRegionManager.swift
import Foundation
import MapboxMaps

// offline region configuration struct
struct OfflineRegion {
    let id: String
    let name: String
    let bounds: CoordinateBounds
    
    // convert CoordinateBounds to Polygon
    var polygon: Polygon {
        let coordinates = [
            CLLocationCoordinate2D(latitude: bounds.southwest.latitude, longitude: bounds.southwest.longitude),
            CLLocationCoordinate2D(latitude: bounds.southwest.latitude, longitude: bounds.northeast.longitude),
            CLLocationCoordinate2D(latitude: bounds.northeast.latitude, longitude: bounds.northeast.longitude),
            CLLocationCoordinate2D(latitude: bounds.northeast.latitude, longitude: bounds.southwest.longitude),
            CLLocationCoordinate2D(latitude: bounds.southwest.latitude, longitude: bounds.southwest.longitude)
        ]
        return Polygon([coordinates])
    }
}

class OfflineRegionManager {

    // ensure the style pack for the Mapbox Standard Style is downloaded
    // the SDK will not re-download it if it's already present
    static func ensureStylePackDownloaded() {
        let offlineManager = OfflineManager()
        
        let stylePackLoadOptions = StylePackLoadOptions(
            glyphsRasterizationMode: .ideographsRasterizedLocally,
            metadata: ["name": "mapbox-standard-stylepack"],
            acceptExpired: false
        )
        
        offlineManager.loadStylePack(
            for: .standard,
            loadOptions: stylePackLoadOptions!
        ) { _ in } completion: { result in
            switch result {
            case let .success(stylePack):
                // Style pack download finishes successfully
                print("Downloaded style pack: \(stylePack)")
            case let .failure(error):
                // Handle error occurred during the style pack download
                if case StylePackError.canceled = error {
                    print("Style pack download cancelled")
                } else {
                    print("Style pack download failed: \(error)")
                }
            }
        }
    }
    
    static func downloadRegion(
        region: OfflineRegion,
        downloadingRegions: Set<String>,
        onDownloadingRegionsUpdate: @escaping (Set<String>) -> Void,
        onProgress: @escaping (String, Float) -> Void,
        onCompletion: @escaping (String, Result<TileRegion, Error>) -> Void
    ) {
        // Don't start if already downloading
        if downloadingRegions.contains(region.id) {
            return
        }
        
        let tileStore = TileStore.default
        let offlineManager = OfflineManager()
        
        // Create tileset descriptor
        let tilesetDescriptor = offlineManager.createTilesetDescriptor(
            for: TilesetDescriptorOptions(
                styleURI: .standard, // get tiles for the Mapbox Standard style
                zoomRange: 10...14,
                tilesets: []
            )
        )
        
        // Create load options using the region's polygon and metadata
        let loadOptions = TileRegionLoadOptions(
            geometry: .polygon(region.polygon),
            descriptors: [tilesetDescriptor],
            metadata: ["name": region.name],
            acceptExpired: true
        )!
        
        // Start downloading - notify UI to add to downloading set
        var updatedDownloadingRegions = downloadingRegions
        updatedDownloadingRegions.insert(region.id)
        onDownloadingRegionsUpdate(updatedDownloadingRegions)
        onProgress(region.id, 0.0)
        
        let _ = tileStore.loadTileRegion(
            forId: region.id,
            loadOptions: loadOptions
        ) { progress in
            let totalResources = max(progress.requiredResourceCount, 1)
            let progressValue = Float(progress.completedResourceCount) / Float(totalResources)
            onProgress(region.id, progressValue)
        } completion: { result in
            // Remove from downloading set and notify completion
            updatedDownloadingRegions.remove(region.id)
            onDownloadingRegionsUpdate(updatedDownloadingRegions)
            onCompletion(region.id, result)
        }
    }
    
    static func clearAllRegions(onCompletion: @escaping () -> Void) {
        let tileStore = TileStore.default
        
        // Get all tile regions from the store
        tileStore.allTileRegions { result in
            switch result {
            case .success(let tileRegions):
                // Remove each region found in the store
                for tileRegion in tileRegions {
                    tileStore.removeRegion(forId: tileRegion.id) { removeResult in
                        switch removeResult {
                        case .success:
                            print("Removed region: \(tileRegion.id)")
                        case .failure(let error):
                            print("Failed to remove region \(tileRegion.id): \(error)")
                        }
                    }
                }
                
                // Clear ambient cache after removing regions
                tileStore.clearAmbientCache { cacheResult in
                    switch cacheResult {
                    case .success(let bytes):
                        print("Cleared \(bytes) bytes from cache")
                    case .failure(let error):
                        print("Failed to clear cache: \(error)")
                    }
                    onCompletion()
                }
                
            case .failure(let error):
                print("Failed to get tile regions: \(error)")
                onCompletion()
            }
        }
    }
}

File: android-offline-maps/app/src/main/java/com/example/mapbox_offline/OfflineRegionManager.kt
package com.example.mapbox_offline

import android.content.Context
import android.util.Log
import com.mapbox.bindgen.Value
import com.mapbox.geojson.Geometry
import com.mapbox.maps.Style
import com.mapbox.common.TileRegion
import com.mapbox.common.TileRegionLoadOptions
import com.mapbox.common.TileStore
import com.mapbox.common.TilesetDescriptor
import com.mapbox.maps.TilesetDescriptorOptions
import com.mapbox.maps.StylePackLoadOptions
import java.util.concurrent.atomic.AtomicReference
import com.mapbox.maps.GlyphsRasterizationMode
import com.mapbox.maps.OfflineManager
import com.mapbox.common.MapboxOptions
import com.mapbox.maps.mapsOptions


object OfflineRegionManager {
    private const val TAG = "OfflineRegionManager"
    private val offlineManager: OfflineManager = OfflineManager()
    private val tileStore: TileStore? = MapboxOptions.mapsOptions.tileStore!!


    fun ensureStylePackDownloaded(context: Context) {
        try {
            val stylePackOptions = StylePackLoadOptions.Builder()
                .glyphsRasterizationMode(GlyphsRasterizationMode.IDEOGRAPHS_RASTERIZED_LOCALLY)
                .metadata(Value("mapbox-standard-stylepack"))
                .acceptExpired(false)
                .build()

            offlineManager.loadStylePack(
                Style.STANDARD,
                stylePackOptions,
                { progress ->
                    //
                },
                { expected ->
                    expected.value?.let { stylePack ->
                        // Style pack download finishes successfully
                        Log.d(TAG, "Style pack downloaded: $stylePack")
                    }
                    expected.error?.let {
                        // Handle error occurred during the style pack download.
                    }
                }
            )

        } catch (e: Throwable) {
            Log.w(TAG, "Style pack API not available or failed: ${e.message}")
        }
    }

    fun downloadRegion(
        region: OfflineRegion,
        downloadingRegions: Set<String>,
        onDownloadingRegionsUpdate: (Set<String>) -> Unit,
        onProgress: (String, Float) -> Unit,
        onCompletion: (String, Result<TileRegion>) -> Unit
    ) {
        if (downloadingRegions.contains(region.id)) {
            Log.d(TAG, "Region ${region.id} is already downloading; skipping")
//            return
        }

        val tilesetDescriptorOptions = TilesetDescriptorOptions.Builder()
            .styleURI(Style.STANDARD)
            .minZoom(10)
            .maxZoom(14)
            .build()

        val tilesetDescriptor = offlineManager.createTilesetDescriptor(tilesetDescriptorOptions)

        

        val loadOptions: TileRegionLoadOptions =TileRegionLoadOptions.Builder()
                .geometry(region.polygon as Geometry)
                .descriptors(listOf(tilesetDescriptor))
                .metadata(Value(region.name))
                .acceptExpired(true)
                .build()
  

        val updated = downloadingRegions.toMutableSet()
        updated.add(region.id)
        onDownloadingRegionsUpdate(updated)
        onProgress(region.id, 0.0f)

        // Ensure tileStore is available before calling its methods
        val ts = tileStore
        if (ts == null) {
            Log.e(TAG, "TileStore not available")
            val updatedAfter = updated.toMutableSet()
            updatedAfter.remove(region.id)
            onDownloadingRegionsUpdate(updatedAfter)
            onCompletion(region.id, Result.failure(Exception("TileStore not available")))
            return
        }
        // Safe to use non-null ts from here on
        ts.loadTileRegion(
             region.id,
             loadOptions,
             { progress ->
                 val total = maxOf(progress.requiredResourceCount, 1)
                 val progressVal = progress.completedResourceCount.toFloat() / total.toFloat()
                 onProgress(region.id, progressVal)
             }
         ) { result ->
             val updatedAfter = updated.toMutableSet()
             updatedAfter.remove(region.id)
             onDownloadingRegionsUpdate(updatedAfter)

             if (result.isValue) {
                 val tileRegion = result.value!!
                 onCompletion(region.id, Result.success(tileRegion))
             } else {
                // result.error may be a Serializable (not a Throwable). Convert safely to Throwable.
                val rawError = result.error
                val throwable = when (rawError) {
                    is Throwable -> rawError
                    null -> Exception("Unknown error")
                    else -> Exception(rawError.toString())
                }
                onCompletion(region.id, Result.failure(throwable))
             }
         }

    }

    fun clearAllRegions(onCompletion: () -> Unit) {
        tileStore?.getAllTileRegions { result ->
            if (result.isValue) {
                val tileRegions = result.value ?: emptyList()
                if (tileRegions.isEmpty()) {
                    tileStore.clearAmbientCache { cacheResult ->
                        if (cacheResult.isValue) {
                            Log.i(TAG, "Cleared ${cacheResult.value} bytes from cache")
                        } else {
                            Log.e(TAG, "Failed to clear cache: ${cacheResult.error}")
                        }
                        onCompletion()
                    }
//                    return
                }

                val removalsPending = AtomicReference(tileRegions.size)
                for (tileRegion in tileRegions) {
                    tileStore.removeTileRegion(tileRegion.id) { removeResult ->
                        if (removeResult.isValue) {
                            Log.i(TAG, "Removed region: ${tileRegion.id}")
                        } else {
                            Log.e(TAG, "Failed to remove region ${tileRegion.id}: ${removeResult.error}")
                        }
                        val remaining = removalsPending.updateAndGet { it - 1 }
                        if (remaining <= 0) {
                            tileStore.clearAmbientCache { cacheResult ->
                                if (cacheResult.isValue) {
                                    Log.i(TAG, "Cleared ${cacheResult.value} bytes from cache")
                                } else {
                                    Log.e(TAG, "Failed to clear cache: ${cacheResult.error}")
                                }
                                onCompletion()
                            }
                        }
                    }
                }
            } else {
                Log.e(TAG, "Failed to get tile regions: ${result.error}")
                onCompletion()
            }
        }
    }
}
